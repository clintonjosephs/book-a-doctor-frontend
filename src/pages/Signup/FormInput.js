import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styles from './FormInput.module.css';

const FormInput = (props) => {
  const [focused, setFocused] = useState(false);

  const {
    errorMessage, onChange, type, name, placeholder, pattern, value,
  } = props;

  const handleFocus = async () => {
    setFocused(true);
  };

  if (pattern === '') {
    return (
      <div className={styles.div_input}>
        <input
          type={type}
          name={name}
          placeholder={placeholder}
          onChange={onChange}
          value={value}
          onBlur={handleFocus}
          focused={focused.toString()}
          onFocus={() => name === 'password_confirmation' && setFocused(true)}
          required
        />
        <span>{errorMessage}</span>
      </div>
    );
  }
  return (
    <div className={styles.div_input}>
      <input
        type={type}
        name={name}
        placeholder={placeholder}
        onChange={onChange}
        value={value}
        onBlur={handleFocus}
        focused={focused.toString()}
        onFocus={() => name === 'password_confirmation' && setFocused(true)}
        required
        pattern={pattern}
      />
      <span>{errorMessage}</span>
    </div>
  );
};

FormInput.propTypes = {
  onChange: PropTypes.func,
  errorMessage: PropTypes.string,
  type: PropTypes.string,
  name: PropTypes.string,
  placeholder: PropTypes.string,
  pattern: PropTypes.string,
  value: PropTypes.string,
};

FormInput.defaultProps = {
  onChange: null,
  errorMessage: '',
  type: '',
  name: '',
  placeholder: '',
  pattern: '',
  value: '',
};

export default FormInput;
