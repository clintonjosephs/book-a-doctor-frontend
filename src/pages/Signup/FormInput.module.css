.div_input span {
  font-size: small;
  align-self: center;
  color: red;
  display: none;
}

.div_input input:invalid[focused='true'] ~ span {
  display: block;
}

.div_input input:invalid[focused='true'] {
  border-color: red;
}
