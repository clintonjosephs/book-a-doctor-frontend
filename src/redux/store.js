import { applyMiddleware, combineReducers, createStore } from 'redux';
import thunk from 'redux-thunk';
import logger from 'redux-logger';
import bookDoctorReducer from './bookDoctor/doctorActions';
import userReducer from './user/userActions';
import reservationReducers from './reservation/reservationActions';

const reducer = combineReducers({
  bookDoctorReducer,
  userReducer,
  reservationReducers,
});

// creating a store
const store = createStore(reducer, applyMiddleware(logger, thunk));

export default store;
