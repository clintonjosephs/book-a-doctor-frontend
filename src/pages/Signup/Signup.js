import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { v4 as uuidv4 } from 'uuid';
import { NavLink, useNavigate } from 'react-router-dom';

import FormInput from './FormInput';
import { signupDispatcher } from '../../redux/signup/signup';

import styles from '../Login/Login.module.css';

const Signup = () => {
  const navigate = useNavigate();

  const dispatch = useDispatch();

  const signupState = useSelector((state) => state.signupReducer.signup || false);
  const errorMessages = useSelector((state) => state.signupReducer.errorMessages || []);

  const [values, setValues] = useState({
    name: '',
    email: '',
    password: '',
    password_confirmation: '',
  });

  const inputs = [
    {
      id: 1,
      name: 'name',
      type: 'text',
      placeholder: 'Name*',
      errorMessage: 'Name should be 3-150 characters.',
      pattern: '^[A-Za-z0-9 ]{3,150}$',
    },
    {
      id: 2,
      name: 'email',
      type: 'email',
      placeholder: 'Email*',
      errorMessage: 'Email should be a valid email.',
    },
    {
      id: 3,
      name: 'password',
      type: 'password',
      placeholder: 'Password*',
      errorMessage: 'Password should be 6-20 characters, without special characters.',
      pattern: '[A-Za-z0-9]{6,20}',
    },
    {
      id: 4,
      name: 'password_confirmation',
      type: 'password',
      placeholder: 'Confirm Password*',
      errorMessage: "Passwords don't match",
      pattern: values.password,
    },
  ];

  const onChange = (e) => {
    setValues({ ...values, [e.target.name]: e.target.value });
  };

  if (signupState) {
    navigate('/');
  }

  const handleSubmit = async (event) => {
    event.preventDefault();

    const data = new FormData();
    data.append('name', values.name);
    data.append('email', values.email);
    data.append('password', values.password);
    data.append('password_confirmation', values.password_confirmation);

    if (event.target.image && event.target.image.files.length !== 0) data.append('image', event.target.image.files[0]);

    dispatch(signupDispatcher(data));
  };

  return (
    <section className={styles.loginContainer}>
      <div className={`${styles.login} ${styles.shadow}`}>
        <h2>Sign Up</h2>
        <div className={styles.line} />
        <div className={styles.failure}>
          {errorMessages.map((element) => (
            <span key={uuidv4()} className="errorMessage">
              {element}
            </span>
          ))}
        </div>

        <form onSubmit={(e) => handleSubmit(e)}>
          {inputs.map((input) => (
            <FormInput
              key={input.id}
              type={input.type}
              name={input.name}
              placeholder={input.placeholder}
              value={values[input.name]}
              onChange={onChange}
              errorMessage={input.errorMessage}
              pattern={input.pattern}
            />
          ))}

          <input type="file" name="image" id="image" />

          <button type="submit" className={styles.formSubmitBtn}>
            Signup
          </button>

          <div style={{ marginTop: '1rem' }}>
            Have an account?
            <NavLink to="/login"> login </NavLink>
          </div>
        </form>
      </div>
    </section>
  );
};

export default Signup;
