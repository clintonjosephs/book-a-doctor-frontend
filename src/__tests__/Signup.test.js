import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

import Signup from '../pages/Signup/Signup';
import FormInput from '../pages/Signup/FormInput';
import store from '../redux/store';

it('Signup renders correctly', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Router>
        <Signup />
      </Router>
    </Provider>,
  );
  expect(tree).toMatchSnapshot();
});

it('FormInput renders correctly', () => {
  const tree = renderer
    .create(
      <Router>
        <FormInput />
      </Router>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
