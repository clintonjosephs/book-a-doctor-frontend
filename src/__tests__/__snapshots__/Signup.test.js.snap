// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput renders correctly 1`] = `
<div
  className="div_input"
>
  <input
    focused="false"
    name=""
    onBlur={[Function]}
    onChange={null}
    onFocus={[Function]}
    placeholder=""
    required={true}
    type=""
    value=""
  />
  <span>
    
  </span>
</div>
`;

exports[`Signup renders correctly 1`] = `
<section
  className="loginContainer"
>
  <div
    className="login shadow"
  >
    <h2>
      Sign Up
    </h2>
    <div
      className="line"
    />
    <div
      className="failure"
    />
    <form
      onSubmit={[Function]}
    >
      <div
        className="div_input"
      >
        <input
          focused="false"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          pattern="^[A-Za-z0-9]{3,150}$"
          placeholder="Username*"
          required={true}
          type="text"
          value=""
        />
        <span>
          Name should be 3-150 characters.
        </span>
      </div>
      <div
        className="div_input"
      >
        <input
          focused="false"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Email*"
          required={true}
          type="email"
          value=""
        />
        <span>
          Email should be a valid email.
        </span>
      </div>
      <div
        className="div_input"
      >
        <input
          focused="false"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          pattern="[A-Za-z0-9]{6,20}"
          placeholder="Password*"
          required={true}
          type="password"
          value=""
        />
        <span>
          Password should be 6-20 characters, without special characters.
        </span>
      </div>
      <div
        className="div_input"
      >
        <input
          focused="false"
          name="password_confirmation"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Confirm Password*"
          required={true}
          type="password"
          value=""
        />
        <span>
          Passwords don't match
        </span>
      </div>
      <input
        id="image"
        name="image"
        type="file"
      />
      <button
        className="formSubmitBtn"
        type="submit"
      >
        Signup
      </button>
      <div
        style={
          Object {
            "marginTop": "1rem",
          }
        }
      >
        Have an account?
        <a
          className=""
          href="/login"
          onClick={[Function]}
        >
           login 
        </a>
      </div>
    </form>
  </div>
</section>
`;
